File: MCRRTImp/src/main/java/lab/mars/MCRRTImp/model/NTreeNode.java
package lab.mars.MCRRTImp.model;
import java.util.*;
import java.util.stream.Collectors;

public class NTreeNode<E> implements Iterable<NTreeNode<E>> {

    private E element;
    private NTreeNode<E> parent;
    private LinkedList<NTreeNode<E>> children;

    public NTreeNode(E element, NTreeNode<E> parent) {
        this.element = element;
        this.children = new LinkedList<>();
        this.parent = parent;
        parent.concatChild(this);
    }

    public NTreeNode(E element) {
        this.element = element;
        this.children = new LinkedList<>();
        this.parent = null;
    }

    @SafeVarargs
    public final void concatChild(NTreeNode<E>... child) {
        for (NTreeNode<E> c : child) {
            c.parent = this;
            this.children.add(c);
        }
    }

    @SafeVarargs
    public final void createChild(E element,E... elements) {
        NTreeNode<E> single = new NTreeNode<>(element);
        single.parent = this;
        children.add(single);
        for (E e : elements) {
            NTreeNode<E> c = new NTreeNode<>(e);
            c.parent = this;
            this.children.add(c);
        }
    }

    public final NTreeNode<E> getChild(int idx) {
        return children.get(idx);
    }

    public final E getElement() {
        return element;
    }

    public final NTreeNode<E> getParent() {
        return parent;
    }

    private class VisitRecorder {
        NTreeNode<E> node;
        int childVisited = 0;

        VisitRecorder(NTreeNode<E> node) {
            this.node = node;
        }
    }

    public interface DistanceFunc<E> {
        double distance(E from, E to);
    }


    public NTreeNode<E> findNearest(E element, DistanceFunc<E> func) {
        NTreeNode<E> minimum = null;
        double minDistance = Double.POSITIVE_INFINITY;
        for (NTreeNode<E> node : this) {
            double dis = func.distance(node.element, element);
            if (dis <= minDistance) {
                minDistance = dis;
                minimum = node;
            }
        }
        return minimum;
    }

    public final List<E> findTrace(E element) {
        Stack<VisitRecorder> stack = new Stack<>();
        stack.push(new VisitRecorder(this));
        while (true) {
            VisitRecorder now = stack.peek();
            if (now.node.children.size() == 0) {
                stack.pop();
                continue;
            }
            if (now.childVisited == now.node.children.size()) {
                stack.pop();
                continue;
            }
            stack.push(new VisitRecorder(now.node.children.get(now.childVisited++)));
            if (stack.peek().node.element.equals(element)) {
                break;
            }
        }
        List<E> trace = stack.stream().map(e -> e.node.element).collect(Collectors.toList());
        return trace;
    }

    public final List<E> findTrace(NTreeNode<E> child) {
        return this.findTrace(child.getElement());
    }

    private NTreeNode<E> self() {
        return this;
    }

    @Override
    public final Iterator<NTreeNode<E>> iterator() {
        return new NTreeIterator();
    }

    public class NTreeIterator implements Iterator<NTreeNode<E>> {


        Stack<VisitRecorder> stack;

        public NTreeIterator() {
            stack = new Stack<>();
            stack.push(new VisitRecorder(self()));
        }

        @Override
        public boolean hasNext() {
            return !stack.empty();
        }

        @Override
        public NTreeNode<E> next() {
            while (!stack.empty()) {
                VisitRecorder now = stack.peek();
                if (now.node.children.size() == 0) {
                    return stack.pop().node;
                }
                if (now.childVisited == now.node.children.size()) {
                    return stack.pop().node;
                }
                stack.push(new VisitRecorder(now.node.children.get(now.childVisited++)));
            }
            throw new RuntimeException("NTreeNode bug");
        }
    }
}

File: MCRRTImp/src/main/kotlin/lab/mars/MCRRTImp/model/NTreeNode.kt
package lab.mars.MCRRTImp.model

import lab.mars.MCRRTImp.base.epsilonEquals
import lab.mars.MCRRTImp.base.random
import java.util.*
import kotlin.collections.ArrayList

class NTreeNode<E>(val element: E) : Iterable<NTreeNode<E>> {

    var parent: NTreeNode<E>? = null

    private val children = LinkedList<NTreeNode<E>>()

    constructor(element: E, parent: NTreeNode<E>) : this(element) {
        this.parent = parent
    }

    fun add(vararg child: NTreeNode<E>) {
        child.forEach { this += it }
    }

    fun add(vararg element: E) {
        element.forEach { this += it }
    }

    operator fun plusAssign(child: NTreeNode<E>) {
        child.parent = this
        this.children.add(child)
    }

    operator fun plusAssign(element: E) {
        this.children.add(NTreeNode(element, this))
    }

    operator fun get(idx: Int): NTreeNode<E> {
        return children[idx]
    }

    val childrenSize
        get() = children.size

    fun forEachChild(func: (child: NTreeNode<E>) -> Unit) {
        children.forEach(func)
    }

    fun clear() {
        children.clear()
    }

    private class VisitRecorder<E>(val node: NTreeNode<E>, var childVisited: Int = 0)

    fun nearestOf(element: E, distanceFunc: (E, E) -> Double): NTreeNode<E> {
        var minimum: NTreeNode<E>? = null
        var minDistance = Double.MAX_VALUE
        this.forEach {
            val distance = distanceFunc(it.element, element)
            if (distance <= minDistance) {
                minDistance = distance
                minimum = it
            }
        }
        return minimum!!
    }

    fun nearestChildOf(element: E, distanceFunc: (NTreeNode<E>, E) -> Double): NTreeNode<E> {
        var minDistance = Double.MAX_VALUE
        val minList = ArrayList<NTreeNode<E>>()
        this.forEach {
            val distance = distanceFunc(it, element)
            if (distance < minDistance) {
                minList.clear()
                minDistance = distance
            }
            if (distance epsilonEquals  minDistance) {
                minList.add(it)
            }
        }
        return minList[(0.0 random minList.size.toDouble()).toInt()]
    }

    fun traceTo(element: E): List<E> {
        val stack = Stack<VisitRecorder<E>>()
        stack.push(VisitRecorder(this))
        while (true) {
            val now = stack.peek()
            if (now.node.children.isEmpty() || now.childVisited == now.node.children.size) {
                stack.pop()
                continue
            }
            val child = now.node.children[now.childVisited++]
            stack.push(VisitRecorder(child))
            if (child.element == element) {
                break
            }
        }
        return stack.map { it.node.element }
    }

    fun traceTo(child: NTreeNode<E>): List<E> {
        return traceTo(child.element)
    }

    override fun iterator(): Iterator<NTreeNode<E>> {
        return NTreeIterator(this)
    }

    private class NTreeIterator<E>(root: NTreeNode<E>) : Iterator<NTreeNode<E>> {

        val stack = Stack<VisitRecorder<E>>()

        init {
            stack.push(VisitRecorder(root))
        }

        override fun hasNext(): Boolean {
            return stack.isNotEmpty()
        }

        override fun next(): NTreeNode<E> {
            while (stack.isNotEmpty()) {
                val now = stack.peek()
                if (now.node.children.isEmpty()) {
                    return stack.pop().node
                }
                if (now.childVisited == now.node.children.size) {
                    return stack.pop().node
                }
                stack.push(VisitRecorder(now.node.children[now.childVisited++]))
            }
            throw RuntimeException("NTreeNode iteration bug")
        }

    }
}